DEV-14638 Working FFI generation for impl items.

Maldron opened the pull request:
Generated modules look like this:
```
pub mod selected_location_selected_field_ffi {
    use super::*;
    use agrian_types::commodity_location::commodity_location_ffi::*;
    use crate::planting::cl_planting_ffi::*;
    pub unsafe extern "C" fn build_commodity_locations(
        selected_field: *const SelectedField,
        plantings: FFIArrayCLPlanting,
    ) -> FFIArrayCommodityLocation {
        let data = (*selected_field).clone();
        let plantings: Vec<CLPlanting> = plantings.into();
        let return_value = data.build_commodity_locations(&*plantings);
        let return_value = match return_value {
            Ok(val) => val,
            Err(error) => {
                ::ffi_common::error::set_last_err_msg(error.to_string().as_str());
                <Vec<CommodityLocation>>::default()
            }
        };
        (&*return_value).into()
    }
}
```

Maldron committed:
DEV-14638 Working FFI generation for impl items.

Generated modules look like this:
```
pub mod selected_location_selected_field_ffi {
    use super::*;
    use agrian_types::commodity_location::commodity_location_ffi::*;
    use crate::planting::cl_planting_ffi::*;
    pub unsafe extern "C" fn build_commodity_locations(
        selected_field: *const SelectedField,
        plantings: FFIArrayCLPlanting,
    ) -> FFIArrayCommodityLocation {
        let data = (*selected_field).clone();
        let plantings: Vec<CLPlanting> = plantings.into();
        let return_value = data.build_commodity_locations(&*plantings);
        let return_value = match return_value {
            Ok(val) => val,
            Err(error) => {
                ::ffi_common::error::set_last_err_msg(error.to_string().as_str());
                <Vec<CommodityLocation>>::default()
            }
        };
        (&*return_value).into()
    }
}
```

Maldron commented on ffi_common/src/macros.rs:
```diff
@@ -56,7 +56,7 @@ the FFI boundary) so we can take care of those steps.
             #[repr(C)]
             #[allow(missing_copy_implementations)]
             #[derive(Clone, Debug)]
-            pub struct [<FFIArray $t:camel>] {
```
Removed the case changes that use `paste` in macros. We need them to match up with the names we generate for the consumer, but we're using `heck` for that since it's String-based, and the case change implementation is slightly different between the two. It wasn't noticeable at first, but when we have something like `CLPlanting`, one will camel-case that as `CLPlanting`, and the other as `Clplanting`.

turboladen replied:
ewww

turboladen commented on ffi_derive/src/lib.rs:
```diff
@@ -200,6 +200,39 @@
+//!
+//! The only caveat here is that the invocation site needs to provide the paths to the FFI modules
+//! that we'll need to import. For example, the FFI form of a function like
+//! `fn meow(&self, volume: Option<Volume>) -> Vec<Meow> { ... }` will need to know the paths to the
```
🥇 🐱

turboladen commented on ffi_derive/src/lib.rs:
```diff
@@ -200,6 +200,39 @@
+//! The only caveat here is that the invocation site needs to provide the paths to the FFI modules
+//! that we'll need to import. For example, the FFI form of a function like
+//! `fn meow(&self, volume: Option<Volume>) -> Vec<Meow> { ... }` will need to know the paths to the
+//! FFI types of `Meow` and `Volume`. Fortunately, since those FFI types probably come from
+//! `ffi_derive`, it's relatively easy to figure out what they would be based on your normal
+//! imports. If you do something like `use crate::animals::cats::Meow;` and
+//! `use utilities::sound::Volume;`, you'll just need to provide
+//! `"crate::animals::cats::meow_ffi", "utilities::sound::volume_ffi"` to the attribute macro.
```
Dunno if this is quite relevant, but maybe worth throwing out there... with declarative macros (aka normal macros), when referencing types or functions from another module or crate, it's conventional to use the full path to thing. Requiring uses to do the same here wouldn't be a stretch IMO.

Maldron replied:
That makes sense, and is probably safer 👍  It feels a little weird to need to know the path of something that you can't see without doing `cargo expand` (or letting rust-analyzer do it for you), but it shouldn't be too weird if it's normal to have to provide the full path for stuff like this.

turboladen commented on ffi_internals/src/impl_internals/impl_ffi.rs:
````diff
@@ -0,0 +1,206 @@
+    /// ```
+    ///
+    pub fn generate_consumer(&self) -> String {
+        format!(
+            r#"
+extension {native_type} {{
+    {functions}
+}}
+            "#,
+            native_type = self.type_name.to_string(),
+            functions = self
+                .fns
+                .iter()
+                .map(|f| f.generate_consumer())
+                .collect::<Vec<String>>()
+                .join("\n"),
+        )
+    }
````
Surely it's non-trival, but my crazy brain just had me wondering if there'd be good ROI on making a `quote!()` that works with Swift code instead of Rust code? Or something along the lines. I suppose that'd entail writing both a parser and a token generator... meh, maybe not.

Maldron replied:
That seems like it'd be super nice to have...lots of work to write, but it would make these easier to write and maintain (especially compared to doing these `format!` templates with multiple languages if we get there...)

turboladen commented on ffi_internals/src/native_type_data.rs:
```diff
@@ -105,6 +114,154 @@ impl From<(NativeType, WrappingType)> for NativeTypeData {
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Option: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Option: {:?}", segment);
+                    }
+                } else if ident == format_ident!("Result") {
+                    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+                        // Take the first generic argument, which is the success type.
+                        if let syn::GenericArgument::Type(arg) = arguments.args.first().unwrap() {
+                            Self::from(UnparsedNativeTypeData {
+                                ty: arg.clone(),
+                                option: unparsed.option,
+                                vec: unparsed.vec,
+                                result: true,
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Result: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Result: {:?}", segment);
+                    }
+                } else if ident == format_ident!("Vec") {
+                    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+                        if let syn::GenericArgument::Type(arg) = arguments.args.first().unwrap() {
+                            Self::from(UnparsedNativeTypeData {
+                                ty: arg.clone(),
+                                option: unparsed.option,
+                                vec: true,
+                                result: unparsed.result,
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Vec: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Vec: {:?}", segment);
+                    }
+                } else {
+                    let native_type = NativeType::from(ident);
```
Is there a nice path for breaking some of this up? Looks like similar shapes in each of those logic gates...

turboladen commented on ffi_internals/src/native_type_data.rs:
```diff
@@ -105,6 +114,154 @@ impl From<(NativeType, WrappingType)> for NativeTypeData {
+    pub fn initial(ty: Type) -> Self {
+        Self {
+            ty,
+            option: false,
+            vec: false,
+            result: false,
```
Would it mess you up to use the more conventional `is_` or `has_` prefix on these?

Maldron replied:
Nope, gc 👍

turboladen reviewed (commented):
From what I can tell, I think the idea here seems good? Since it's still in Draft, I'm assuming you care more about the big picture, but I left you some comments on the details anyway. :)

Maldron committed:
Clean some things up and write a working example for tests. Still need to add the actual test, but it expands properly.

turboladen commented on ffi_derive/tests/trait_impl.rs:
```diff
@@ -0,0 +1,45 @@
+
+mod animals {
+    pub mod cats {
```
lol

Maldron commented on ffi_derive/src/lib.rs:
```diff
@@ -200,6 +200,39 @@
+//! use crate::animals::cats::Meow;
+//! use utilities::sound::Volume;
+//! #[derive(ffi_derive::expose_items(
+//!     "crate::animals::cats::meow_ffi",
+//!     "utilities::sound::volume_ffi"
```
Nope! I just keep forgetting that paths can be passed like this. It's so much nicer a) using paths and b) having the full path passed in. You even get warning highlights on the macro invocation if you give it something that isn't used in the generated code.

Maldron commented on ffi_internals/src/impl_internals/impl_ffi.rs:
```diff
@@ -0,0 +1,206 @@
+                                syn::FnArg::Receiver(_receiver) => acc.1 = true,
+                                syn::FnArg::Typed(arg) => {
+                                    let argument_name = if let syn::Pat::Ident(pat) = &*arg.pat {
+                                        pat.ident.clone()
+                                    } else {
+                                        panic!(
+                                            "Anonymous parameter (not allowed in Rust 2018): {:?}",
+                                            input
+                                        );
+                                    };
+                                    let native_type_data = NativeTypeData::from(
+                                        UnparsedNativeTypeData::initial(*arg.ty.clone()),
+                                    );
+                                    acc.0.push(FnParameterFFI {
+                                        name: argument_name,
+                                        native_type_data,
+                                        original_type: *arg.ty.clone(),
+                                    })
+                                }
+                            }
+                            acc
+                        },
+                    );
+
+                    let return_type: Option<NativeTypeData> = match &method.sig.output {
+                        syn::ReturnType::Default => None,
+                        syn::ReturnType::Type(_token, ty) => Some(NativeTypeData::from(
+                            UnparsedNativeTypeData::initial(*ty.clone()),
+                        )),
+                    };
+
+                    Some(FnFFI {
+                        fn_name,
+                        has_receiver,
+                        parameters: arguments,
+                        return_type,
+                    })
+                } else {
+                    None
+                }
```
Yeah, split this up into `From` impls for `FnFFI` and `FnParameterFFI`.

Maldron committed:
DEV-14638 Finish test and refactor.

Add a test that verifies the behavior of the generated FFI.

Clean up the repeated patterns and stop using strings when paths are a
real type.
